### PR TITLE
Better handling of ASTDelayedPredicate

### PR DESCRIPTION
**Pull request by apmoriarty**
Addresses https://github.com/NationalSecurityAgency/datawave/issues/1270

### CODE REVIEW

**Review by FineAndDandy — DISMISSED** [review]

**Review by hlgp — DISMISSED** [review]

File: warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java [hunk 1]
```diff
@@ -44,10 +44,18 @@ public String toString() {
...
      * @param node
-     * @return
+     *            a JexlNode
+     * @return an ASTDelayedPredicate, or the original node if this node is already delayed
      */
-    public static ASTDelayedPredicate create(JexlNode node) {
+    public static JexlNode create(JexlNode node) {
+        
+        // Do not delay if this subtree is already delayed
+        if (QueryPropertyMarker.findInstance(node).isType(ASTDelayedPredicate.class) || ASTDelayedPredicate.isSubTreeAlreadyDelayed(node)) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java]: It kind of looks like this `QueryPropertyMarker.findInstance(node).isType(ASTDelayedPredicate.class)` might already be handled if you just did `if (isSubTreeAlreadyDelayed(node))`. Not super used to query code though. Is that right? Or could the first check be true when the other check is false?

File: warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java [hunk 2]
```diff
@@ -59,4 +67,48 @@ public static ASTDelayedPredicate create(JexlNode node) {
...
+    
+    /**
+     * Unwrap a delayed predicate, fully. Intended to handle the odd edge case when multiple delayed predicate markers are applied to the same node
+     *
+     * @param node
+     *            an arbitrary jexl node
+     * @return the source node, or the original node if this node is not delayed
+     */
+    public static JexlNode unwrapFully(JexlNode node) {
+        if (QueryPropertyMarker.findInstance(node).isType(ASTDelayedPredicate.class)) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java]: Not sure how often we'll be doing this or how deep these trees can grow but you could save a traversal by doing:

```
Instance instance = QueryPropertyMarker.findInstance(node);
if (instance.isType(ASTDelayedPredicate.class)) {
  JexlNode source = instance.getSource();
  // ...
```

File: warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java]: I'm leaning towards closing this PR and working these cases into https://github.com/NationalSecurityAgency/datawave/pull/1297

File: warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java [hunk 3]
```diff
@@ -59,4 +67,48 @@ public static ASTDelayedPredicate create(JexlNode node) {
...
+            
+            QueryPropertyMarker.Instance instance = QueryPropertyMarker.findInstance(node);
+            JexlNode source = instance.getSource();
+            
+            while (QueryPropertyMarker.findInstance(source).isType(ASTDelayedPredicate.class)) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java]: Similar duplicate traversals are happening here:

```
while (instance.isType(ASTDelayedPredicate.class)) {
  instance = QueryPropertyMarker.findInstance(instance.getSource());
}
```

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java]: Great catch, done!

**Comment by apmoriarty** [comment]
I prefer https://github.com/NationalSecurityAgency/datawave/pull/1386 to this merge request and am in favor of closing this one

**Comment by apmoriarty** [comment]
Closing now that https://github.com/NationalSecurityAgency/datawave/pull/1386 is merged

**Pull request closed without merging**